lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -541,6 +541,85 @@ static void getCellAsmResultNames(OpAsmSetValueNameFn setNameFn, Operation *op,
     setNameFn(op->getResult(i), prefix + portNames[i].str());
 }
 
+//===----------------------------------------------------------------------===//
+// AssignOp
+//===----------------------------------------------------------------------===//
+
+/// Returns whether the given value is a port of a component, which are
+/// defined as Block Arguments.
+static bool isComponentPort(Value value) {
+  return value.getImpl()->getKind() == detail::ValueImpl::Kind::BlockArgument;
+}
+
+/// Verifies the given value of the AssignOp if it is a component port. The
+/// `isDestination` boolean is used to distinguish whether the value is a source
+/// or a destination.
+static LogicalResult verifyAssignOpWithComponentPort(AssignOp op,
+                                                     bool isDestination) {
+  Value value = isDestination ? op.dest() : op.src();
+
+  auto component = op->getParentOfType<ComponentOp>();
+  Block *body = component.getBody();
+  auto it = llvm::find_if(body->getArguments(),
+                          [&](auto arg) { return arg == value; });
+  BlockArgument blockArg = *it;
+  size_t blockArgNum = blockArg.getArgNumber();
+
+  auto ports = getComponentPortInfo(component);
+  assert(blockArgNum < ports.size() &&
+         "Block argument index should be within range of component's ports.");
+  // Within a component, we want to drive component output ports, and be
+  // driven by its input ports.
+  Direction validComponentDirection = isDestination ? Output : Input;
+  return ports[blockArgNum].direction == validComponentDirection
+             ? success()
+             : op.emitOpError()
+                   << "has a component port as the "
+                   << (isDestination ? "destination" : "source")
+                   << " with the incorrect direction. It should be "
+                   << (isDestination ? "Output" : "Input") << ".";
+}
+
+/// Verifies the destination of an assignment operation.
+static LogicalResult verifyAssignOpDestination(AssignOp assign) {
+  Value dest = assign.dest();
+  if (isComponentPort(dest))
+    return verifyAssignOpWithComponentPort(assign,
+                                           /*isDestination=*/true);
+
+  Operation *definingOp = dest.getDefiningOp();
```

Comment:
Maybe factor out this logic to an `bool isDriveable(Value value)` function

Reply:
Originally I did something like this, but I couldn't come up with a clean way to only include the Value. If it is a component port, it won't have a defining operation; it will instead be a BlockArgument. For this reason, I factored out the `verifyAssignOpWithComponentPort`.

Reply:
If there is some way to get the component from just the `Value`, that I'm missing, that would fix this.